File: WebApp/indoor-localization/src/data-access/mock-repositories/MockAssetPositionHistoryLogRepository.ts
import { AssetPositionHistoryLog } from "../../entities/AssetPositionHistoryLog";
import { AssetPositionHistoryLogFilter } from "../../entities/AssetPositionHistoryLogFilter";
import { IAssetPositionHistoryLogRepository } from "../repository-interfaces/IAssetPositionHistoryLogRepository";

export class MockAssetPositionHistoryLogRepository implements IAssetPositionHistoryLogRepository {
    private static logs: AssetPositionHistoryLog[] = [];
  
    async Get(id: number): Promise<AssetPositionHistoryLog | null> {
      return MockAssetPositionHistoryLogRepository.logs.find(log => log.id === id) || null;
    }

    async GetLogs(assetPositionHistoryLogFilter: AssetPositionHistoryLogFilter): Promise<AssetPositionHistoryLog[]> {
        return MockAssetPositionHistoryLogRepository.logs.filter(log => {
          const isFacilityMatch = assetPositionHistoryLogFilter.facilityId
            ? log.facility.id === assetPositionHistoryLogFilter.facilityId
            : true;
          const isAssetMatch = assetPositionHistoryLogFilter.assetId
            ? log.asset.id === assetPositionHistoryLogFilter.assetId
            : true;
          const isDateInRange = log.dateTime >= assetPositionHistoryLogFilter.startDate && log.dateTime <= assetPositionHistoryLogFilter.endDate;
    
          return isFacilityMatch && isAssetMatch && isDateInRange;
        });
      }
  
    async GetAll(): Promise<AssetPositionHistoryLog[]> {
      return MockAssetPositionHistoryLogRepository.logs;
    }
  
    async Add(log: AssetPositionHistoryLog): Promise<boolean> {
      MockAssetPositionHistoryLogRepository.logs.push(log);
      return true;
    }
  
    async Delete(id: number): Promise<boolean> {
      const index = MockAssetPositionHistoryLogRepository.logs.findIndex(log => log.id === id);
      if (index !== -1) {
        MockAssetPositionHistoryLogRepository.logs.splice(index, 1);
        return true;
      }
      return false;
    }
  
    async Update(updatedLog: AssetPositionHistoryLog): Promise<boolean> {
      const index = MockAssetPositionHistoryLogRepository.logs.findIndex(log => log.id === updatedLog.id);
      if (index !== -1) {
        MockAssetPositionHistoryLogRepository.logs[index] = updatedLog;
        return true;
      }
      return false;
    }
  }
  

File: Android/indoor_localization/lib/domain/mock-repositories/mock_asset_position_history_log_repository.dart
import 'package:indoor_localization/domain/entities/asset_position_history_log.dart';
import 'package:indoor_localization/domain/entities/asset_position_history_log_filter.dart';
import 'package:indoor_localization/domain/repository-interfaces/asset_position_history_log_repository.dart';

class MockAssetPositionHistoryLogRepository implements IAssetPositionHistoryLogRepository {
  static List<AssetPositionHistoryLog> logs = [];

  @override
  AssetPositionHistoryLog? get(int id) {
    return logs.firstWhere((log) => log.id == id);
  }

  @override
  List<AssetPositionHistoryLog> getLogs(AssetPositionHistoryLogFilter filter) {
    return logs.where((log) {
      final isFacilityMatch = filter.facilityId != null
          ? log.facility.id == filter.facilityId
          : true;
      final isAssetMatch = filter.assetId != null
          ? log.asset.id == filter.assetId
          : true;
      final isDateInRange = log.dateTime.isAfter(filter.startDate) &&
          log.dateTime.isBefore(filter.endDate);

      return isFacilityMatch && isAssetMatch && isDateInRange;
    }).toList();
  }

  @override
  List<AssetPositionHistoryLog> getAll() {
    return logs;
  }

  @override
  bool add(AssetPositionHistoryLog log) {
    logs.add(log);
    return true;
  }

  @override
  bool delete(int id) {
    final index = logs.indexWhere((log) => log.id == id);
    if (index != -1) {
      logs.removeAt(index);
      return true;
    }
    return false;
  }

  @override
  bool update(AssetPositionHistoryLog updatedLog) {
    final index = logs.indexWhere((log) => log.id == updatedLog.id);
    if (index != -1) {
      logs[index] = updatedLog;
      return true;
    }
    return false;
  }
}
